### ISSUE TITLE
/mark returns 504 timeout error

### ISSUE THREAD

**Issue by msandbothe** [CarnegieLearningWeb/UpGrade#7]
Noticed that locust script was failing on every call to /mark. I attempted to recreate this via curl commands and am hitting the 1 minute timeout on /mark. Of note /assign is returning fine in the command line, but producing an error in the UI

INIT
`curl -X 'POST' \
  '$/api/init' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "Authorization: Bearer $(< token.txt)" \
  -d '{
  "userId": "test123",
  "group": {
  	"schoolId": ["testschool"],
  	"classId": ["testclass"],
  	"instructorId": ["testteacher"]},
  "workingGroup": {
  	"schoolId": ["testschool"],
  	"classId": ["testclass"],
  	"instructorId": ["testteacher"]}
}'`

Result:
`{
   "userId": "test123",
   "group": {
      "schoolId": [
         "testschool"
      ],
      "classId": [
         "testclass"
      ],
      "instructorId": [
         "testteacher"
      ]
   },
   "workingGroup": {
      "schoolId": [
         "testschool"
      ],
      "classId": [
         "testclass"
      ],
      "instructorId": [
         "testteacher"
      ]
   },
   "id": "69bafeb6-041c-46d9-8a35-c847e3f39372",
   "createdAt": "2021-04-28T20:09:30.879Z",
   "updatedAt": "2021-04-28T20:09:30.879Z",
   "versionNumber": 1
}`

ASSIGN
`curl -X 'POST' \
  '$/api/assign' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "Authorization: Bearer $(< token.txt)" \
  -d '{
  "userId": "test123",
  "context": "assign-prog"
}'`

Result:
`[
   {
      "expId": "absolute_value_plot_equality",
      "expPoint": "SelectSection",
      "twoCharacterId": "N7",
      "assignedCondition": {
         "createdAt": "2021-04-07T18:12:25.844Z",
         "updatedAt": "2021-04-07T18:12:25.844Z",
         "versionNumber": 1,
         "id": "db2482dc-f7e9-43d1-9a81-3260df0659de",
         "twoCharacterId": "KQ",
         "name": "",
         "description": "Second Condition",
         "conditionCode": "variant",
         "assignmentWeight": 50
      }
   },
   {
      "expId": "simplify_exponents_2_power",
      "expPoint": "SelectSection",
      "twoCharacterId": "QX",
      "assignedCondition": {
         "conditionCode": "default"
      }
   },
   {
      "expId": "volume_surface_area_sphere_vol",
      "expPoint": "SelectSection",
      "twoCharacterId": "S2",
      "assignedCondition": {
         "conditionCode": "default"
      }
   },
   {
      "expId": "volume_surface_area_cone_vol",
      "expPoint": "SelectSection",
      "twoCharacterId": "RU",
      "assignedCondition": {
         "conditionCode": "default"
      }
   }
]`

Frontend UI:
`Experiment user not defined occurred in accessing $/api/assign view
4:09 PM

Group not defined for experiment User: {
  "createdAt": "2021-04-28T17:57:27.431Z",
  "updatedAt": "2021-04-28T17:57:27.431Z",
  "versionNumber": 1,
  "id": "test123",
  "group": null,
  "workingGroup": null,
  "experiment": [
    "2bcb1aca-e410-4e01-ab27-86278c3f00b0",
    "27dbab64-48d0-4308-9323-b74bd8c90ebc"
  ]
}`

MARK
`curl -X 'POST' \
  '$/api/mark' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "Authorization: Bearer $(< token.txt)" \
  -d '{
  "userId": "test123",
  "experimentPoint": "SelectSection",
  "partitionId": "absolute_value_plot_equality",
  "condition": "variant"
}'`

Result:
1 minute timeout

**Comment by ppratikcr7** [comment]
Hi @msandbothe. I tried to call mark from postman and it was a success, I was able to get enrolment count increased. Please see the below screenshot. I figured out that the current locust script fails on /mark call with timeout error is due to the functionality set in that script. Actually the script first asks to delete one or all experiments. Then it asks to create 'n' experiments and based on the created experiments the condition, experimentPoint and expId pairs present in the newly created experiment through the locust script are stored in a list. We later use a random selected pair of condition, experimentPoint and expId to mark an experiment. So, if we create a manual experiment from the UI the corresponding pairs wont be available in the locust script as of now and so we cant mark those pairs. If you want, you can have the values of your AA1-3 experiments JSON in the script instead of dynamically generated data and it should work fine.

<img width="1271" alt="Screenshot 2021-04-30 at 5 56 39 PM" src="https://user-images.githubusercontent.com/33730817/116695773-ac17b200-a9de-11eb-8715-5f83da116362.png">

**Comment by msandbothe** [comment]
I went through my curl commands again and I think I got it working, but I think the API might be confusing me. The results from /assign give me `{"expId":"volume_surface_area_cone_vol","expPoint":"SelectSection","twoCharacterId":"RU","assignedCondition":{"conditionCode":"default"}}`

If I plug that in to the /mark command as I would expect (`"experimentPoint": "SelectSection", "partitionId": "volume_surface_area_cone_vol", "condition": "default"`), I get this error `{"name":"Error","message":"Condition not found: default","endPoint":"/api/mark","errorCode":null,"type":"Condition not found","createdAt":"2021-05-04T14:22:26.159Z","updatedAt":"2021-05-04T14:22:26.159Z","versionNumber":1,"id":252482}
`
If I do not provide the condition (`"experimentPoint": "SelectSection", "partitionId": "volume_surface_area_cone_vol"`) I get this error `{"name":"Error","message":"Condition not found: undefined","endPoint":"/api/mark","errorCode":null,"type":"Condition not found","createdAt":"2021-05-04T14:23:24.591Z","updatedAt":"2021-05-04T14:23:24.591Z","versionNumber":1,"id":252483}
`
If I explicitly send the condition as null (`"experimentPoint": "SelectSection", "partitionId": "volume_surface_area_cone_vol", "condition": null`)  I get a successful result `{"createdAt":"2021-05-04T14:23:51.508Z","updatedAt":"2021-05-04T14:23:51.508Z","versionNumber":1,"id":"volume_surface_area_cone_vol_SelectSection_thisisnotarealuser","experimentId":"volume_surface_area_cone_vol_SelectSection","enrollmentCode":"Student included in experiment","userId":"thisisnotarealuser","condition":null} `

**Comment by msandbothe** [comment]
#39 tracks the issues regarding conditions returned from /assign. This may or may not be the root cause of the 504 timeout mentioned by this issue.

**Comment by ppratikcr7** [comment]
Okay. Let me check the above issue #39 as discussed on slack.

**Comment by ppratikcr7** [comment]
@msandbothe  I have responded back on issue #39, we can now close issue #7 and #39 as both seemed to have some format corrections in the curl requests.

**Comment by msandbothe** [comment]
Still investigating, even after correcting the group key issue found in #39, the AA1 experiment /mark call still fails to return.

**Comment by msandbothe** [comment]
I created a shallow copy of AA1 on a different site ID, First attempt at /mark succeeded.

One hypothesis I thought of is that AA1 v1 has 295 students enrolled (experiment should at 500). Is it possible that another number is being compared to 500 and saying "no more enrollments", and causing mark to fail silently perhaps. This could be related to #13 in that case.

**Comment by ppratikcr7** [comment]
Hi @msandbothe , I tried to check your hypothesis on AA1 experiment. I have set the enrollment end criteria to “1 Student” to quickly check your hypothesis. I enrolled one student successfully without any /mark call 504 error. After that it shows the “Enrollment Complete” Status as seen in below screenshot. Now, I tried to enroll another student where I call the assign api and I am getting the response in which the conditionCode is “default”. After this also my mark call didn’t gave 504 error and it executed properly. So, this is not related to issue #13.
